### PR TITLE
TRM-22548 Update the disease service to use BasicSearchController and…

### PR DESCRIPTION
**Pull request by sahmad-ebi**
Changes in disease controller to use basecontroller and basesevice,

### CODE REVIEW

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseController.java [hunk 1]
```diff
@@ -3,47 +3,72 @@
...
+import java.util.Optional;
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+import static uk.ac.ebi.uniprot.api.rest.output.UniProtMediaType.*;
 
 @RestController
 @RequestMapping("/disease")
 @Validated
-public class DiseaseController {
+public class DiseaseController extends BasicSearchController<Disease> {
```
Comment (LeonardoGonzales) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseController.java]: I did not find the Download method, will you implement it as part of another task?

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseJsonMessageConverter.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import uk.ac.ebi.uniprot.api.rest.output.converter.AbstractJsonMessageConverter;
+import uk.ac.ebi.uniprot.cv.disease.Disease;
+import uk.ac.ebi.uniprot.json.parser.disease.DiseaseJsonConfig;
+import uk.ac.ebi.uniprot.search.field.DiseaseField;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+
+public class DiseaseJsonMessageConverter extends AbstractJsonMessageConverter<Disease> {
```
Comment (LeonardoGonzales) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseJsonMessageConverter.java]: I think this class is no longer required. once you make the AbstractJsonMessageConverter not abstract anymore

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseTsvMessageConverter.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import uk.ac.ebi.uniprot.parser.tsv.disease.DiseaseEntryMap;
+import uk.ac.ebi.uniprot.search.field.DiseaseField;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DiseaseTsvMessageConverter extends AbstractTsvMessagerConverter<Disease> {
+    public static final String COMMA = "\\s*,\\s*";
```
Comment (LeonardoGonzales) [inline support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseTsvMessageConverter.java]: Can you please add it to an Util class

File: common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/AbstractJsonMessageConverter.java [hunk 4]
```diff
@@ -24,12 +22,20 @@
     private final ObjectMapper objectMapper;
     private ThreadLocal<Map<String, List<String>>> tlFilters = new ThreadLocal<>();
     private ThreadLocal<JsonGenerator> tlJsonGenerator = new ThreadLocal<>();
+    private List<ReturnField> allFields;
+    private JsonResponseFieldProjector fieldProjector;
 
-    public AbstractJsonMessageConverter(ObjectMapper objectMapper, Class<T> messageConverterEntryClass) {
+    public AbstractJsonMessageConverter(ObjectMapper objectMapper, Class<T> messageConverterEntryClass) { // TODO Remove this constructor once all converter use the other constructor
```
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/AbstractJsonMessageConverter.java]: As of now, this constructor is being used by many converter such as KeywordJsonMessageConverter, UniProtKBJsonMessage Converter. I will file a separate ticket to fix this.

[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/AbstractJsonMessageConverter.java]: Yes. that is good point. I will rename the class once the dependency on filterEntryContent is removed from all the classes.

File: common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/JsonResponseFieldProjector.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    // get the value of a given field
+    private Object getFieldValue(Object source, String fieldName) {
+        try {
+            Field sourceField = source.getClass().getDeclaredField(fieldName);
+            sourceField.setAccessible(true);
+            return sourceField.get(source);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            // incorrect fieldName provided
+            return null;
```
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/JsonResponseFieldProjector.java]: Added.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java [hunk 6]
```diff
@@ -1,72 +1,43 @@
...
-import java.util.stream.Collectors;
+import uk.ac.ebi.uniprot.search.field.DiseaseField;
 
 @Service
 public class DiseaseService {
-    private static final String ACCESSION_STR = "accession:";
-    @Autowired
-    private DiseaseRepository diseaseRepository;
+    private BasicSearchService<Disease, DiseaseDocument> basicService;
+    private DefaultSearchHandler defaultSearchHandler;
```
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java]: I didn't understand the idField in DefaultSearchHandler constructor. When i set the id as name. My search was not working.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseController.java (hunk 1 above)
Comment (sahmad-ebi) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseController.java]: There is a separate ticket for download.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseTsvMessageConverter.java (hunk 3 above)
Comment (sahmad-ebi) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/response/converter/DiseaseTsvMessageConverter.java]: Added.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java [hunk 7]
```diff
@@ -1,72 +1,43 @@
...
+
     @Autowired
     private DiseaseSolrSortClause solrSortClause;
+
     @Autowired
-    private DiseaseDocumentToDiseaseConverter toDiseaseConverter;
+    public void setDefaultSearchHandler() {
+        //this.defaultSearchHandler = new DefaultSearchHandler(DiseaseField.Search.content,
```
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java]: fixed it.

File: support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply support-data-rest/src/main/java/uk/ac/ebi/uniprot/api/disease/DiseaseService.java]: fixed it.

File: common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/JsonResponseFieldProjector.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Returns a map of root level fields with their values for the json writer
+ */
+public class JsonResponseFieldProjector {
+
+    // Get a map containing the object's projected fields
+    public Map<String, Object> project(Object object, List<String> returnFields, List<ReturnField> allFields) {
+
+        if(!Utils.notEmpty(returnFields)){ // if the return fields is empty return all fields from allFields
+            returnFields = allFields.stream().map(f -> f.toString()).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (sahmad-ebi) [reply common-rest/src/main/java/uk/ac/ebi/uniprot/api/rest/output/converter/JsonResponseFieldProjector.java]: As per java doc
Most programmers should use the toString() method in preference to the name() method, as the toString() method may return a more user-friendly name.

**Pull request merged**